alf/utils/losses.py

```diff
@@ -102,3 +104,371 @@ def multi_quantile_huber_loss(quantiles: torch.Tensor,
+        Returns:
+            loss with the same shape as target
+        """
+        raise NotImplementedError()
+
+    def calc_expectation(self, pred: torch.Tensor):
+        """Calculate the expected predition in the untransfomred domain from ``pred``.
+        """
+        raise NotImplementedError()
+
+    def initialize_bias(self, bias: torch.Tensor):
+        """Initialize the bias of the last FC layer for the prediction properly.
+
+        Args:
+            bias: the bias parameter to be initialized.
+        """
+        with torch.no_grad():
+            bias.zero_()
+
+
+class SquareLoss(ScalarPredictionLoss):
+    """Square loss for predicting scalar target.
+
+    Args:
+        transform: the transformation applied to target. If it is provided, the
+            the regression target will be transformed.
+    """
+
+    def __init__(self, transform: Optional[InvertibleTransform] = None):
+        super().__init__()
+        self._transform = transform
+
+    def __call__(self, pred: torch.Tensor, target: torch.Tensor):
+        """Calculate the loss.
+
+        Args:
+            pred: shape is [B]
+            target: the shape is [B]
+        Returns:
+            loss with the same shape as target
+        """
+        assert pred.shape == target.shape
+
+        if self._transform is not None:
+            target = self._transform.transform(target)
+        return (pred - target)**2
+
+    def calc_expectation(self, pred: torch.Tensor):
+        """Calculate the expected predition in the untransfomred domain from ``pred``.
+
+        Args:
+            pred: raw model prediction
+        """
+        if self._transform is not None:
+            pred = self._transform.inverse_transform(pred)
+        return pred
+
+
+def _get_indexer(shape):
+    ndim = len(shape)
+    ones = [1] * ndim
+    B = tuple(
+        torch.arange(d).reshape(d, *ones[i + 1:]) for i, d in enumerate(shape))
+    return B
+
+
+class _DiscreteRegressionLossBase(ScalarPredictionLoss):
+    """The base class for DiscreteRegressionLoss and OrderedDiscreteRegresionLoss."""
+
+    def __init__(self,
+                 transform: Optional[InvertibleTransform] = None,
+                 inverse_after_mean=False):
+        super().__init__()
+        self._transform = transform
+        if self._transform is not None:
+            self._inverse_after_mean = inverse_after_mean
+        else:
+            self._inverse_after_mean = True
+        self._support = None
+
+    def _calc_support(self, n):
+        if self._support is not None and self._support.shape[0] == n:
+            return self._support
+        upper_bound = n // 2
+        lower_bound = -((n - 1) // 2)
+        x = torch.arange(lower_bound, upper_bound + 1, dtype=torch.float32)
+        if self._transform is not None and not self._inverse_after_mean:
+            x = self._transform.inverse_transform(x)
+        self._support = x
+        return x
+
+    def _calc_bin(self, logits, target):
```

Comment:
Suggesting adding documentation for this core function.

Reply:
added.